File: ext/ts/test/golden.test.js
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { test } from "node:test";

import {
  CapCommands,
  CapTools,
  decodeHeader,
  decodeHello,
  decodeInterceptReq,
  decodeSubscribe,
  encodeFrame,
  encodeHello,
  encodeInterceptReq,
  EvToolCall,
} from "../dist/pxb/index.js";

const here = dirname(fileURLToPath(import.meta.url));
const goldenDir = join(here, "../../go/pxb/testdata");

test("decode Go golden hello.bin", () => {
  const raw = readFileSync(join(goldenDir, "hello.bin"));
  const h = decodeHello(raw);
  assert.equal(h.name, "greet");
  assert.equal(h.version, "1.0.0");
  assert.equal(h.caps, CapTools | CapCommands);
  assert.equal(h.protocol, 1);
});

test("decode Go golden intercept_req.bin", () => {
  const raw = readFileSync(join(goldenDir, "intercept_req.bin"));
  const ix = decodeInterceptReq(raw);
  assert.equal(ix.event, EvToolCall);
  assert.equal(ix.toolName, "bash");
  assert.equal(ix.toolCallID, "c1");
  assert.equal(Buffer.from(ix.input).toString(), '{"command":"ls"}');
});

test("decode Go golden subscribe.bin", () => {
  const raw = readFileSync(join(goldenDir, "subscribe.bin"));
  const s = decodeSubscribe(raw);
  assert.deepEqual(s.events, [5, 10]);
  assert.deepEqual(s.intercept, [1]);
});

test("decode Go golden hello_frame.bin", () => {
  const raw = readFileSync(join(goldenDir, "hello_frame.bin"));
  const hdr = decodeHeader(raw.subarray(0, 16));
  assert.equal(hdr.type, 1);
  assert.equal(hdr.payload, raw.length - 16);
  const h = decodeHello(raw.subarray(16));
  assert.equal(h.name, "greet");
});

test("TS encode matches Go golden hello payload", () => {
  const got = encodeHello({
    name: "greet",
    version: "1.0.0",
    caps: CapTools | CapCommands,
    protocol: 1,
  });
  const want = readFileSync(join(goldenDir, "hello.bin"));
  assert.deepEqual(Buffer.from(got), want);
});

test("TS encodeInterceptReq round-trips", () => {
  const body = encodeInterceptReq({
    event: EvToolCall,
    toolName: "bash",
    toolCallID: "c1",
    input: Buffer.from('{"command":"ls"}'),
    content: "",
    isError: false,
    errText: "",
    prompt: "",
    reason: "",
    targetID: "",
    turnIndex: 0,
  });
  const want = readFileSync(join(goldenDir, "intercept_req.bin"));
  assert.deepEqual(Buffer.from(body), want);
});

test("encodeFrame header+body", () => {
  const body = encodeHello({ name: "x", version: "1", caps: 0, protocol: 1 });
  const frame = encodeFrame(1, 0, 0, body);
  const hdr = decodeHeader(frame.subarray(0, 16));
  assert.equal(hdr.type, 1);
  assert.equal(hdr.payload, body.length);
});

File: ext/go/pxb/golden_test.go
package pxb_test

import (
	"os"
	"path/filepath"
	"slices"
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/pulseaiclub/phi/ext/pxb"
)

// TestWriteGoldenFixtures regenerates ext/go/pxb/testdata/*.bin for TS interop.
// Run with UPDATE_GOLDEN=1 (or pass -update) to rewrite fixtures.
func TestWriteGoldenFixtures(t *testing.T) {
	if os.Getenv("UPDATE_GOLDEN") != "1" && !updateFlag() {
		t.Skip("set UPDATE_GOLDEN=1 or -update to regenerate")
	}
	dir := "testdata"
	require.NoError(t, os.MkdirAll(dir, 0o755))

	write := func(name string, body []byte) {
		path := filepath.Join(dir, name)
		require.NoError(t, os.WriteFile(path, body, 0o644))
	}

	hello := pxb.EncodeHello(pxb.Hello{
		Name: "greet", Version: "1.0.0", Caps: pxb.CapTools | pxb.CapCommands, Protocol: 1,
	})
	write("hello.bin", hello)

	ack := pxb.EncodeHelloAck(pxb.HelloAck{
		Protocol: 1, PhiVersion: "v0.19.0", Cwd: "/tmp", SessionID: "s1", ExtensionDir: "/ext",
	})
	write("hello_ack.bin", ack)

	ix := pxb.EncodeInterceptReq(pxb.InterceptReq{
		Event: pxb.EvToolCall, ToolName: "bash", ToolCallID: "c1",
		Input: []byte(`{"command":"ls"}`),
	})
	write("intercept_req.bin", ix)

	sub := pxb.EncodeSubscribe(pxb.Subscribe{
		Events: []uint16{pxb.EvSessionStart, pxb.EvAgentEnd}, Intercept: []uint16{pxb.EvToolCall},
	})
	write("subscribe.bin", sub)

	frame := append([]byte(nil), make([]byte, pxb.HeaderSize)...)
	pxb.EncodeHeader(frame, pxb.Header{Type: pxb.TypeHello, Flags: 0, ID: 0, Payload: uint32(len(hello))})
	frame = append(frame[:pxb.HeaderSize], hello...)
	write("hello_frame.bin", frame)
}

func updateFlag() bool {
	return slices.Contains(os.Args, "-update")
}

func TestGoldenFixturesDecode(t *testing.T) {
	dir := "testdata"
	helloRaw, err := os.ReadFile(filepath.Join(dir, "hello.bin"))
	if err != nil {
		t.Skip("testdata missing; run UPDATE_GOLDEN=1 go test ./pxb -run TestWriteGoldenFixtures")
	}
	h, err := pxb.DecodeHello(helloRaw)
	require.NoError(t, err)
	require.Equal(t, "greet", h.Name)
	require.Equal(t, pxb.CapTools|pxb.CapCommands, h.Caps)

	ixRaw, err := os.ReadFile(filepath.Join(dir, "intercept_req.bin"))
	require.NoError(t, err)
	ix, err := pxb.DecodeInterceptReq(ixRaw)
	require.NoError(t, err)
	require.Equal(t, "bash", ix.ToolName)
	require.JSONEq(t, `{"command":"ls"}`, string(ix.Input))
}
